### ISSUE TITLE
Rendering engine used by Roundcube

### ISSUE THREAD

**Issue by ngarnier** [roundcube/roundcubemail#5761]
Hello, 

After some research, I can't find any information on the rendering engine used by Roundcube to render an HTML email (my bad if it's out there, checked the forum and Github).

I'm especially interested as Roundcube seems to be one of the last email clients in which we can't get [fully clickable buttons](https://github.com/mjmlio/mjml/pull/668#issuecomment-301502350) to work, for [MJML](https://mjml.io/).

Thanks a lot!

**Comment by alecpl** [comment]
It uses custom code, partially based in Washtml library. What exactly problem do you have? What is the example HTML code you want to support?

**Comment by ngarnier** [comment]
Thanks for the answer @alecpl. I attached an `.html` file containing an example of the clickable button.

The approach is to wrap the whole button (based on a `table` layout because... email) inside an `a` tag, and also have an `a` tag as the last HTML node wrapped inside a conditional comment for Outlook. 

In Roundcube, the button is not at all clickable (not on the actual button but also not even on the text inside the button). 

![](http://g.recordit.co/O2iwJhbZTA.gif)

**Comment by alecpl** [comment]
So, under the hood we use PHP's DOMDocument parser (i.e. libxml). I'm not yet sure where's the issue, but it looks that:
1. Everything inside `<!--[if mso | IE]>` is remove. It's not an issue here.
2. `<a><table>...</table></a>` becomes `<a></a><table>...</table>` which is the main problem.

**Comment by alecpl** [comment]
While we could fix 1. by handling the HTML comments before using DOMDocument parser we can do nothing about 2.

**Comment by alecpl** [comment]
I said we could, but I'm not really sure we should. That would be IE-only and not so simple task. https://msdn.microsoft.com/en-us/library/ms537512(v=vs.85).aspx#syntax Patches welcome.

**Comment by alecpl** [comment]
For 2. it might be reasonable to evaluate possibility to replace our current parser with https://github.com/Masterminds/html5-php. My quick test proved your button works with this parser, but a couple of unit tests fail (nothing serious) and I do not know yet if it handles all corner cases and what is performance or security impact.

**Comment by ngarnier** [comment]
I agree with you about 1., I don't think Roundcube should handle those comments which are really used for Microsoft Outlook by the email community. It also doesn't look like a clean solution.

It would be amazing to see another parser implemented if it means better support for HTML and more opportunities for email, although I do understand that is this some serious change and needs to be done carefully. 

I'll follow this closely!

**Comment by alecpl** [comment]
Here's the patch to test Masterminds/HTML5 parser:
```patch
diff --git a/composer.json-dist b/composer.json-dist
index 465d2e8..7b4fbfd 100644
--- a/composer.json-dist
+++ b/composer.json-dist
@@ -23,6 +23,7 @@
         "pear/crypt_gpg": "~1.6.0",
         "pear/net_sieve": "~1.4.0",
         "roundcube/plugin-installer": "~0.1.6",
+        "masterminds/html5": "~2.2.2",
         "endroid/qrcode": "~1.6.5"
     },
     "require-dev": {
diff --git a/program/lib/Roundcube/rcube_washtml.php b/program/lib/Roundcube/rcube_washtml.php
index b45a747..c95bde8 100644
--- a/program/lib/Roundcube/rcube_washtml.php
+++ b/program/lib/Roundcube/rcube_washtml.php
@@ -536,8 +536,12 @@ class rcube_washtml
         $method       = $this->is_xml ? 'loadXML' : 'loadHTML';
         $options      = 0;
.
+        if (class_exists('Masterminds\HTML5')) {
+            $html5 = new Masterminds\HTML5();
+            $node  = $html5->loadHTML($html);
+        }
         // Use optimizations if supported
-        if (PHP_VERSION_ID >= 50400) {
+        else if (PHP_VERSION_ID >= 50400) {
             $options = LIBXML_PARSEHUGE | LIBXML_COMPACT | LIBXML_NONET;
             @$node->{$method}($html, $options);
         }
```

**Comment by CRtEurope** [comment]
I must say this patch is working great so far

**Comment by alecpl** [comment]
@CRtEurope did you run it for some longer time? Do you have more feedback on this?
@thomascube what do you think?

The library development is not very active, but also I don't see many bugs in the tracker. I think we could use it and if we get many complaints we could make it optional dependency.
